### ISSUE TITLE
[FEATURE] Support CPU Pinning

### ISSUE THREAD

**Issue by janeczku** [harvester/harvester#2305]
Certain VM workloads, requiring a predictable latency and enhanced performance during its execution would benefit from obtaining dedicated CPU resources. KubeVirt, relying on the Kubernetes CPU manager, is able to pin guest's vCPUs to the host's pCPUs.
Harvester should expose this functionality to the user.

Ref: https://kubevirt.io/user-guide/virtual_machines/dedicated_cpu_resources/

**Comment by gusraggio** [comment]
This is needed in telco use case. Also related to this but different is CPU prioritization some VMs, need to be prioritized to others but dont need a CPU assigned to a VM.

For example, if the server is overloaded prioritized VMs should be respected. Do we need a separeted issue?

**Comment by torchiaf** [comment]
@brandboat I think Harvester server should return the new `enableCPUManager` and `disableCPUManager` nodes actions along with the already available actions in the `nodes` API response:

![image](https://github.com/user-attachments/assets/94e6d616-e462-4af4-92b8-e23b0b2443f7)

so that the UI can show the action buttons accordingly with current mechanism.
Should we also update the [HEP](https://github.com/harvester/harvester/pull/5803) as well ?
cc @bk201

**Comment by brandboat** [comment]
> @brandboat I think Harvester server should return the new enableCPUManager and disableCPUManager nodes actions along with the already available actions in the nodes API response:
> 
> so that the UI can show the action buttons accordingly with current mechanism.
> Should we also update the https://github.com/harvester/harvester/pull/5803 as well ?

Thanks @torchiaf, the enableCPUManager and disableCPUManager apis are mentioned in  https://github.com/harvester/harvester/blob/master/enhancements/20240506-cpu-pinning.md#api-changes, about the API response, I'll update the HEP after implementation done.

**Comment by TachunLin** [comment]
Related GUI issue: https://github.com/harvester/harvester/issues/6056

**Comment by TachunLin** [comment]
### Related Test Plan

1. Test enable/disable cpu manager through API on Harvester
2. Test enable/disable cpu pinning through API on virtual machine
3. Test the cpu pinning vm can request specific core cpu on host to it's vgpu.
4. Verify only one master node could enable/disable cpu manager in the same time 
5. Verify there should be only one job enable/disable cpu manager per node (API only)
6. Verify we can't enable cpu manager when it is enabled, and we can't disable CPU manager when it's disabled (API only)
7. Test reboot node with cpu manager enabled
    - Node should back to Active
    - Harvester cluster is working 
    - VM enabled cpu pinning should be running and request dedicate cpu core on node
8. Test reboot vm 
    - CPU pinning configuration exists
    - vm can restarted well 
    - vm request dedicate cpu core on node
    - After reboot VM will be located on any node with cpu manager enabled  
9. Test migrate vm with cpu pinning 
    - Can migrate vm to the node with cpu manager enabled
    - Filter out the node with no cpu manager enabled (UI)
    - Non cpu-pinning vm migrate to cpu-manager enabled node

10. Disable the node have cpu manager when w/ and w/o cpu pinning VM on it.
 

11. Test upgrade path (master to `v1.4.0-rc1`)
     - Scope:  Single, multiple nodes upgrade  
     - Can complete finish the upgrade process
     - After upgrade complete, the cpu pinning vm should work as expected

Notice during the test
1. Consider cpu overcommit won't take effect on resource overcommit (request, limit),CPU pinning config will overwrite cpu overcommit
2. (Negative) During upgrade, vm migrate to the node not enabled cpu manager
3. Take care the availale resource on the node with cpu manager enabled, it would more likely out of resource to create new vm on it.

**Comment by TachunLin** [comment]
Verified **fixed** on `v1.4-cd10d18-head` (24/09/19). Close this issue

### Result
<details><summary>$\color{green}{\textsf{PASS}}$ Test enable/disable cpu manager through API on Harvester $~~$</summary>

  1. Can correctly enable `CPU manager` through API on Harvester node


  ```
  davidtclin@localhost:~/Documents/Harvester/workspace/cypress/cypress-140-string-update-fix/harvester-tests/cypress> curl 'https://192.168.0.131/v1/harvester/nodes/harvester-node-0?action=enableCPUManager' \
  > -X 'POST' \
  > -H 'Authorization: Bearer token-f5nxx:tqsq77xbchvs8wrclnrddfw62gzmvzrj95vkrl94rqm444s7hzmcjs' \
  > --insecure
  ```
  ```
  harvester-node-0:~ # kubectl get nodes harvester-node-0 -o yaml | grep cpumanager
      cpumanager: "true"
  ```
  
  ![image](https://github.com/user-attachments/assets/53d0d150-6d06-4775-ab81-edabdbe58fcb)
  
  ![image](https://github.com/user-attachments/assets/dedf804b-4950-4d35-be5b-be4c8f54bbcc)


  
  2. Can correctly disable `CPU manager` through API on Harvester node

  ```
  davidtclin@localhost:~/Documents/Harvester/workspace/cypress/cypress-140-string-update-fix/harvester-tests/cypress> curl 'https://192.168.0.131/v1/harvester/nodes/harvester-node-0?action=disableCPUManager' -X 'POST' -H 'Authorization: Bearer token-f5nxx:tqsq77xbchvs8wrclnrddfw62gzmvzrj95vkrl94rqm444s7hzmcjs' --insecure
  ```

```
harvester-node-0:~ # kubectl get nodes harvester-node-0 -o yaml | grep cpumanager
    cpumanager: "false"
```

  ![image](https://github.com/user-attachments/assets/11cb88e2-95b4-4dfa-8518-3a611f5f5748)



 </details>


<details><summary>$\color{green}{\textsf{PASS}}$ Test enable/disable cpu pinning through API on virtual machine $~~$</summary>

  1. Can correctly enable `CPU pinning` through API on virtual machine
  
  ```
  harvester-node-0:~ # kubectl get virtualmachine test -n default -o yaml | grep dedicatedCpuPlacement 
            dedicatedCpuPlacement: true
  ```

2. CPU pinning vm have the correct `vcpupin` requests in the virsh xml content

  ```
  harvester-node-0:~ # kubectl exec virt-launcher-test-xpds5 -- virsh dumpxml default_test | awk "/<cputune>/,/<\/cputune>/"
    <cputune>
      <vcpupin vcpu='0' cpuset='1'/>
      <vcpupin vcpu='1' cpuset='2'/>
    </cputune>
  ```
  

  3. User can't disable cpu manager when there is a cpu pinned vm in harvester-node-0

  ```
  admission webhook "validator.harvesterhci.io" denied the request: current cpu manager policy is already the same as requested value: staticdavidtclin@localhosodes/harvester-node-2?action=disableCPUManager' -X 'POST' -H 'Authorization: Bearer token-tldrj:9gfmrhr78qlxx28tr28xqxzlzwj76nvq4ckmp54nvcjg8wm495z887' --insecure
  admission webhook "validator.harvesterhci.io" denied the request: there should not be any running VMs with CPU pinning when disabling the CPU manager
  ```

  

  4. Stop the vm, then we can disable the cpu manager on harvester-node-2 (host the cpu pinning) vm

  ![image](https://github.com/user-attachments/assets/b871dea4-e694-407f-8ff5-af547ce5588f)

  5. Can correctly disable `CPU pining` through API on virtual machine   

  ![image](https://github.com/user-attachments/assets/2153eadf-dd0e-4175-ac29-d3fcbf03fd0f)

  6. **After restart** the Non CPU pinning vm did not have `vcpupin` requests in the virsh xml content 

  ```
  harvester-node-0:~ # kubectl exec virt-launcher-test-t5m8g -- virsh dumpxml default_test | awk "/<cputune>/,/<\/cputune>/"
  ```

 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ Test the cpu pinning vm can request specific core cpu on host to it's vgpu  $~~$</summary>


  1. Can correctly enable `CPU pinning` through API on virtual machine
  
  ```
  harvester-node-0:~ # kubectl get virtualmachine test -n default -o yaml | grep dedicatedCpuPlacement 
            dedicatedCpuPlacement: true
  ```

2. CPU pinning vm have the correct `vcpupin` requests in the virsh xml content

  ```
  harvester-node-0:~ # kubectl exec virt-launcher-test-xpds5 -- virsh dumpxml default_test | awk "/<cputune>/,/<\/cputune>/"
    <cputune>
      <vcpupin vcpu='0' cpuset='1'/>
      <vcpupin vcpu='1' cpuset='2'/>
    </cputune>
  ```
  

 </details>


<details><summary>$\color{green}{\textsf{PASS}}$ Verify only one master node could enable/disable cpu manager in the same time  $~~$</summary>


  1. Can't enable cpu manager on another node when there is a node still in updating.   
  
  ```
  davidtclin@localhost:~/Documents/Harvester> curl 'https://192.168.0.131/v1/harvester/nodes/harvester-node-0?action=enableCPUManager' -X 'POST' -H 'Authorization: Bearer token-tldrj:9gfmrhr78qlxx28tr28xqxzlzwj76nvq4ckmp54nvcjg8wm495z887' --insecure
  davidtclin@localhost:~/Documents/Harvester> curl 'https://192.168.0.131/v1/harvester/nodes/harvester-node-1?action=enableCPUManager' -X 'POST' -H 'Authorization: Bearer token-tldrj:9gfmrhr78qlxx28tr28xqxzlzwj76nvq4ckmp54nvcjg8wm495z887' --insecure
  admission webhook "validator.harvesterhci.io" denied the request: the node you are trying to update the cpu manager policy is a master node, and only one master node can be updated at a time, while job update-cpu-manager-harvester-node-0-gr9hj is updating the policy for other master nodes
  ```
  2. Can't disable cpu manager on another node when there is a node still in updating. 
  ```
  davidtclin@localhost:~/Documents/Harvester> curl 'https://192.168.0.131/v1/harvester/nodes/harvester-node-1?action=disableCPUManager' -X 'POST' -H 'Authorization: Bearer token-tldrj:9gfmrhr78qlxx28tr28xqxzlzwj76nvq4ckmp54nvcjg8wm495z887' --insecure
  davidtclin@localhost:~/Documents/Harvester> curl 'https://192.168.0.131/v1/harvester/nodes/harvester-node-0?action=disableCPUManager' -X 'POST' -H 'Authorization: Bearer token-tldrj:9gfmrhr78qlxx28tr28xqxzlzwj76nvq4ckmp54nvcjg8wm495z887' --insecure
  admission webhook "validator.harvesterhci.io" denied the request: the node you are trying to update the cpu manager policy is a master node, and only one master node can be updated at a time, while job update-cpu-manager-harvester-node-1-lhzsh is updating the policy for other master nodes
  ```

 </details>


<details><summary>$\color{green}{\textsf{PASS}}$ Verify there should be only one job enable/disable cpu manager per node (API only)  $~~$</summary>


  1. Can't enable cpu manager on the same node when the current node still in updating.   

  ```
  davidtclin@localhost:~/Documents/Harvester> curl 'https://192.168.0.131/v1/harvester/nodes/harvester-node-0?action=enableCPUManager' -X 'POST' -H 'Authorization: Bearer token-tldrj:9gfmrhr78qlxx28tr28xqxzlzwj76nvq4ckmp54nvcjg8wm495z887' --insecure
  davidtclin@localhost:~/Documents/Harvester> curl 'https://192.168.0.131/v1/harvester/nodes/harvester-node-0?action=enableCPUManager' -X 'POST' -H 'Authorization: Bearer token-tldrj:9gfmrhr78qlxx28tr28xqxzlzwj76nvq4ckmp54nvcjg8wm495z887' --insecure
  admission webhook "validator.harvesterhci.io" denied the request: there is other job update-cpu-manager-harvester-node-0-rhpkc updating the cpu manager policy for this node harvester-node-0
  ```

  2. Can't disable cpu manager on the same node when the current node still in updating. 

  ```
  davidtclin@localhost:~/Documents/Harvester> curl 'https://192.168.0.131/v1/harvester/nodes/harvester-node-1?action=disableCPUManager' -X 'POST' -H 'Authorization: Bearer token-tldrj:9gfmrhr78qlxx28tr28xqxzlzwj76nvq4ckmp54nvcjg8wm495z887' --insecure
  davidtclin@localhost:~/Documents/Harvester> curl 'https://192.168.0.131/v1/harvester/nodes/harvester-node-1?action=disableCPUManager' -X 'POST' -H 'Authorization: Bearer token-tldrj:9gfmrhr78qlxx28tr28xqxzlzwj76nvq4ckmp54nvcjg8wm495z887' --insecure
  admission webhook "validator.harvesterhci.io" denied the request: there is other job update-cpu-manager-harvester-node-1-zg725 updating the cpu manager policy for this node harvester-node-1
  ```

 </details>


<details><summary>$\color{green}{\textsf{PASS}}$ Verify we can't enable/disable cpu manager when it is enabled/disabled and (API only) $~~$</summary>


  1. Can't enable cpu manager when it's already enabled on that node

  ```
  davidtclin@localhost:~/Documents/Harvester> curl 'https://192.168.0.131/v1/harvester/nodes/harvester-node-0?action=enableCPUManager' -X 'POST' -H 'Authorization: Bearer token-tldrj:9gfmrhr78qlxx28tr28xqxzlzwj76nvq4ckmp54nvcjg8wm495z887' --insecure
  admission webhook "validator.harvesterhci.io" denied the request: current cpu manager policy is already the same as requested value: static
  ```

  2. Can't disable cpu manager when it's already disabled on that node

  ```
  davidtclin@localhost:~/Documents/Harvester> curl 'https://192.168.0.131/v1/harvester/nodes/harvester-node-0?action=disableCPUManager' -X 'POST' -H 'Authorization: Bearer token-tldrj:9gfmrhr78qlxx28tr28xqxzlzwj76nvq4ckmp54nvcjg8wm495z887' --insecure
  admission webhook "validator.harvesterhci.io" denied the request: current cpu manager policy is already the same as requested value: none
  ```

 </details>


<details><summary>$\color{green}{\textsf{PASS}}$ Test reboot node with cpu manager enabled  $~~$</summary>


  1. After reboot node, the node should back to Active

![image](https://github.com/user-attachments/assets/34cf5d4d-7b46-4271-a102-a32a280e809d)


  2. The Harvester cluster is working

  ```
  harvester-node-0:~ # kubectl get nodes
  NAME               STATUS   ROLES                       AGE   VERSION
  harvester-node-0   Ready    control-plane,etcd,master   11d   v1.28.12+rke2r1
  harvester-node-1   Ready    control-plane,etcd,master   11d   v1.28.12+rke2r1
  harvester-node-2   Ready    control-plane,etcd,master   11d   v1.28.12+rke2r1
  ```

  3.  CPU manager is still enabled on that node 

  ```
  harvester-node-0:~ # kubectl get nodes harvester-node-0 -o yaml | grep cpumanager
      cpumanager: "true"
  ```
  
  4. VM enabled cpu pinning should be running well 

  ![image](https://github.com/user-attachments/assets/004862e0-90fe-4729-94fb-49c031e21168)

  
  5. VM request dedicate cpu core on node

  ```
  harvester-node-0:~ # kubectl exec virt-launcher-test0-hwhq7 -- virsh dumpxml default_test0 | awk "/<cputune>/,/<\/cputune>/"
    <cputune>
      <vcpupin vcpu='0' cpuset='1'/>
      <vcpupin vcpu='1' cpuset='2'/>
    </cputune>
  ```

 </details>


<details><summary>$\color{green}{\textsf{PASS}}$ Test reboot cpu pinned vm  $~~$</summary>


  1. After reboot vm, the CPU pinning configuration exists

  ```
  harvester-node-0:~ # kubectl get virtualmachines test0 -n default -o yaml | grep dedicate
            dedicatedCpuPlacement: true
  ```

  2. VM can restarted well into running 

  ![image](https://github.com/user-attachments/assets/f8d09bb1-4630-49d4-92e7-aefdcf2d3b67)


  3.  vm request dedicate cpu core on node

  ```
  harvester-node-0:~ # kubectl exec virt-launcher-test0-qcpm8 -- virsh dumpxml default_test0 | awk "/<cputune>/,/<\/cputune>/"
    <cputune>
      <vcpupin vcpu='0' cpuset='1'/>
      <vcpupin vcpu='1' cpuset='2'/>
    </cputune>
  ```

  4. After reboot VM will be located on any node with cpu manager enabled
      Restarted vm 10 times, most of the time, vm will started on harvester-node-1

  ![image](https://github.com/user-attachments/assets/f0eb1482-bfbd-4bc5-a8f9-8279f7e951c3)


 </details>


<details><summary>$\color{green}{\textsf{PASS}}$ Test migrate vm with cpu pinning  $~~$</summary>


  1. While migrating cpu pinning vm, can display all nodes have cpu manager enabled 
  ![image](https://github.com/user-attachments/assets/5dad81e6-7f98-420c-bf4d-f2d46138a930)
  ![image](https://github.com/user-attachments/assets/2f815af3-25da-46f1-be35-2fa2c6ee77af)
  
  
  2. Can migrate vm to the node with cpu manager enabled

  ![image](https://github.com/user-attachments/assets/555bcda2-099d-4d49-b995-9ad9da1c77bf)


  3. VM can restarted well into running 

  ![image](https://github.com/user-attachments/assets/8ef2695b-e25d-4b55-87a7-18ac85b8c4a8)


  4.  Filter out the node with no cpu manager enabled (UI)

  ![image](https://github.com/user-attachments/assets/7642cfb2-daa6-4f39-bd8c-9f41fc1928a9)
  ![image](https://github.com/user-attachments/assets/6a5517b2-a6ae-4093-9ac9-b6c50b22de1b)


  5. Non cpu-pinning vm can migrate to cpu-manager enabled node with the correct configuration

  ![image](https://github.com/user-attachments/assets/4eef2e7b-7bf2-4332-bb56-57eba881031a)
  ![image](https://github.com/user-attachments/assets/7f1d660b-b922-4ea6-835b-027180792a8a)
  ![image](https://github.com/user-attachments/assets/4ce7de28-6d5a-4e34-bb92-1f597197367c)


 </details>


<details><summary>$\color{green}{\textsf{PASS}}$ Disable the node have cpu manager when with cpu pinning VM on it $~~$</summary>


  1. Disable the node have cpu manager when with cpu pinning vm on it

![image](https://github.com/user-attachments/assets/1064f457-209d-4767-8203-15354c7c67c0)

  ```
  davidtclin@localhost:~/Documents/Harvester> curl 'https://192.168.0.131/v1/harvester/nodes/harvester-node-0?action=disableCPUManager' -X 'POST' -H 'Authorization: Bearer token-tldrj:9gfmrhr78qlxx28tr28xqxzlzwj76nvq4ckmp54nvcjg8wm495z887' --insecure
  admission webhook "validator.harvesterhci.io" denied the request: there should not be any running VMs with CPU pinning when disabling the CPU manager
  ```
  
  2. After stop the running vm, we can disable the cpu manager on that node 

  ![image](https://github.com/user-attachments/assets/25905908-3d98-4c3c-983e-d8cc09e4710e)


  3. Then we can start the cpu pinning vm well


  4. And the cpu pinning vm will be started on the node have cpu manager enabled 

  ![image](https://github.com/user-attachments/assets/ae3b4ba8-8de1-4061-87b2-5917f3ee31cf)


 </details>


<details><summary>$\color{green}{\textsf{PASS}}$ <strong>(Single node)</strong> Test upgrade path (v1.4.0-rc1 to v1.4.0-rc1) $~~$</summary>


1.  Can correctly upgrade to v1.4.0-rc1 with cpu manager enabled and have cpu pinning vm
  ![image](https://github.com/user-attachments/assets/9324b9e4-1ae9-4e6c-865b-6ac46d0b8896)

2. After upgrade, the cpu manager option is still enabled on the node 
  ![image](https://github.com/user-attachments/assets/4e99f299-48b2-40b5-b53d-80a3e956f71c)

3. We can start the cpu pinning vm, it still enabled the cpu pinning option
  ![image](https://github.com/user-attachments/assets/31a84e4e-0911-4cc0-b414-3edb0e0bf5ee)

4. vm still request dedicate cpu core on node

  ```
  harv140rc1:~ # kubectl exec virt-launcher-test-xphlk -- virsh dumpxml default_test | awk "/<cputune>/,/<\/cputune>/"
    <cputune>
      <vcpupin vcpu='0' cpuset='1'/>
      <vcpupin vcpu='1' cpuset='2'/>
    </cputune>
  ```

5. Can disable the cpu manager on node


 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ <strong>(Multiple nodes)</strong> Test upgrade path (v1.4.0-rc1 to v1.4.0-rc1) $~~$</summary>


1.  Can correctly upgrade to v1.4.0-rc1 with cpu manager enabled and have cpu pinning vm

  ![image](https://github.com/user-attachments/assets/4f1ee21f-152e-4336-9f4c-643ba20d9f05)


2. After upgrade, the cpu manager option is still enabled on the node 

  ![image](https://github.com/user-attachments/assets/69946696-b402-4c45-b474-eca11a85e434)



3. The cpu pinning vm, it running and enabled the cpu pinning option

  ![image](https://github.com/user-attachments/assets/57a8f445-837a-4280-be35-0ea0af4f0b21)


4. vm still request dedicate cpu core on node

  ```
  harvester-node-2:~ # kubectl exec virt-launcher-test1-q6c4f -- virsh dumpxml default_test1 | awk "/<cputune>/,/<\/cputune>/"
    <cputune>
      <vcpupin vcpu='0' cpuset='1'/>
      <vcpupin vcpu='1' cpuset='2'/>
    </cputune>
  harvester-node-2:~ # k9s
  harvester-node-2:~ # kubectl exec virt-launcher-test-lb45m -- virsh dumpxml default_test | awk "/<cputune>/,/<\/cputune>/"
    <cputune>
      <vcpupin vcpu='0' cpuset='1'/>
      <vcpupin vcpu='1' cpuset='2'/>
    </cputune>
  ```

5. Can disable the cpu manager on node

  ![image](https://github.com/user-attachments/assets/0f912319-974e-43de-b3f1-2e91785f2113)


 </details>


### Test Information
* Test Environment: Three nodes kvm machines
* Harvester version: `v1.4-cd10d18-head` (24/09/19)

### Verify Steps
<details><summary> Test enable/disable cpu manager through API on Harvester </summary>

  1. Create api key: click dashboard top right profile picture > Account & API Keys > Create API Key 
  2. Use curl cmd to enable cpu manager on `harvester-node-0`
  ```
      curl 'https://192.168.100.131/v1/harvester/nodes/{node_name}?action=enableCPUManager' \
      -X 'POST' \
      -H 'Authorization: Bearer <API_KEY>' \
      --insecure
  ```  
  4. Check the cpu manager have enabled on the selected node on UI 
  6. Check the cpu manager have enabled on the selected node from node content 
   ```
     kubectl get nodes harvester-node-0 -o yaml | grep cpumanager
   ```  
  7. Use curl command to disable CPU manger on harvester-node-0 
  ```
      curl 'https://192.168.100.131/v1/harvester/nodes/{node_name}?action=enableCPUManager' \
      -X 'POST' \
      -H 'Authorization: Bearer <API_KEY>' \
      --insecure
  ```
  8. Check the cpu manager have disabled on the selected node on UI 
  9. Check the cpu manager have disabled on the selected node from node content 
  


</details>

<details><summary> Test enable/disable cpu pinning through API on virtual machine </summary>

  1. Already enable cpu manager on the node ready to host the vm 
  2. Create a vm with 2 core cpu and 2Gi memory
  3. Edit the yaml
  5. Add `dedicatedCpuPlacement: true` to `.spec.template.spec.domain.cpu` 
  ![image](https://github.com/user-attachments/assets/baafe425-a5b2-482d-9dd7-d99992378c5b)
  
  6. Create the vm
  7. Access the Harvester node
  8. Check the virt-launcher pod of the vm 
  10. Access the vm-launcher pod to get the virsh xml content of the vm 
  11. Check the vm have requested the 2 dedicated core CPU from node 
    ```
    kubectl exec virt-launcher-test-t5m8g -- virsh dumpxml default_test | awk "/<cputune>/,/<\/cputune>/"
    ```
  12. Disable the cpu manager on the node have cpu pinning vm 
  13. Check should be blocked with message
  14. Stop the vm 
  15. Disable the cpu manager on the node again 
  16. Check can disable the cpu manager correctly 
  17. Start the vm
  18. Remove the `dedicatedCpuPlacement: true` to `.spec.template.spec.domain.cpu`
  19. Restart the vm
  20.  Check the vm did not request any core CPU from node

</details>

<details><summary> Verify only one master node could enable/disable cpu manager in the same time </summary>

  1. Ensure non of the node have enable the cpu manger 
  2. Use curl command to enable cpu manager on the first node harvester-node-0 
  3. Don't wait for the update complete
  4. Directly enable cpu manager on the second node harvester-node-1
  6. Check should be blocked with error message
  7. Wait for the update complete
  8. Enable cup manager on the second node 
  9. After node-0 and node-1 already enabled cpu manager
  10. Use curl command to disable cpu manager on the first node harvester-node-1 
  11. Don't wait for the update complete
  12. Directly disable cpu manager on the second node harvester-node-1
  13. Check should be blocked with error message

</details>

<details><summary> Verify there should be only one job enable/disable cpu manager per node (API only) </summary>

  1. Ensure non of the node have enable the cpu manger 
  2. Use curl command to enable cpu manager on the first node harvester-node-0 
  3. Wait for the update complete
  4. Enable cpu manager on the first node harvester-node-0
  5. Check should be blocked with error message
  6. Disable cpu manager on node-0
  7. Wait for the update complete
  9. Use curl command to disable cpu manager on the first node harvester-node-0
  10. Check should be blocked with error message

</details>

<details><summary> Verify we can't enable/disable cpu manager when it is enabled/disabled and (API only) </summary>

  1. Ensure non of the node have enable the cpu manger 
  2. Use curl command to enable cpu manager on the first node harvester-node-0 
  3. Don't wait for the update complete
  4. Directly enable cpu manager on the second node harvester-node-0
  5. Check should be blocked with error message
  6. Enable cup manager on node-1
  7. After node-1 already enabled cpu manag
  8. Use curl command to disable cpu manager on the first node harvester-node-1 
  9. Don't wait for the update complete
  11. Directly disable cpu manager on the second node harvester-node-1
  12. Check should be blocked with error message

</details>

<details><summary> Test reboot node with cpu manager enabled  </summary>

  1. Harvester-node-0 have cpu manager enabled and have cpu pinning vm on it
  2. Reboot node0 
  3. Check after reboot node, the node should back to Active
  4. Check the Harvester cluster is working
  5. Check the cpu manager is still enabled on that node
  6. Check VM enabled cpu pinning should be running well
  7. Check VM can request dedicate cpu core on node


</details>

<details><summary> Test reboot cpu pinned vm </summary>

  1. Harvester-node-0 have cpu manager enabled and have cpu pinning vm on it
  2. Reboot the vm
  3. Check after reboot vm, the CPU pinning configuration exists
  4. Check VM can restarted well into running 
  5. Check vm can request dedicate cpu core on node
  6. Check after reboot VM will be located on any node with cpu manager enabled


</details>

<details><summary> Test migrate vm with cpu pinning </summary>

  1. Harvester-node-0 have cpu manager enabled and have cpu pinning vm on it 
  2. Harvester-node-1 and node-2 also enable cpu manager 
  3. Migrate the vm 
  4. Check all nodes with cpu manager enabled can display on the migration target list 
  5. Check can migrate vm to the node with cpu manager enabled
  6. Check the VM can restarted well into running
  7. Disable the cpu manager on node-1 node
  8. Migrate the vm 
  9. Check the target list do not display the node-1 
  10. Create another vm with no cpu pinning named `no-pinning`
  11. Migrate the `no-pinning` vm
  12. Check can display all the nodes on the target list
  13. Check the migrate the vm and running well 
  14. Check the migrated vm did not have cpu pinning enabled.




</details>

<details><summary> Disable the node have cpu manager when with cpu pinning VM on it 
 </summary>

  1. Harvester-node-0 have cpu manager enabled and have cpu pinning vm on it 
  2. Disable the cpu manager when vm is running 
  3. Check should display the block message
  4. Stop the vm
  5. Disable the cpu manager again 
  6. Check can disable the cpu manager on node-0
  7. Start the vm
  8. Check the vm will be started on another node with cpu manager enabled

</details>
